File: Client/file-exploder/Models/QueueJob.swift
import Foundation

struct QueueJob: Identifiable, Codable {
    let id: String
    let type: OperationType
    let srcPath: String?
    let dstPath: String?
    let mode: String?
    let status: JobStatus
    let error: String?
    let createdAt: Date
    let startedAt: Date?
    let completedAt: Date?

    /// The daemon serialises jobs with snake_case keys (see `queue.Job` in the
    /// Go server), so the mapping has to be spelled out.
    enum CodingKeys: String, CodingKey {
        case id
        case type
        case srcPath = "src_path"
        case dstPath = "dst_path"
        case mode
        case status
        case error
        case createdAt = "created_at"
        case startedAt = "started_at"
        case completedAt = "completed_at"
    }

    enum OperationType: String, Codable {
        case rename
        case move
        case delete
        case copy
        case mkdir
        case chmod
        
        var displayName: String {
            switch self {
            case .rename: return "名前変更"
            case .move: return "移動"
            case .delete: return "削除"
            case .copy: return "コピー"
            case .mkdir: return "新規フォルダ"
            case .chmod: return "権限変更"
            }
        }
        
        var systemImage: String {
            switch self {
            case .rename: return "pencil"
            case .move: return "arrow.right.doc.on.clipboard"
            case .delete: return "trash"
            case .copy: return "doc.on.doc"
            case .mkdir: return "folder.badge.plus"
            case .chmod: return "lock.shield"
            }
        }
    }
    
    enum JobStatus: String, Codable {
        case pending
        case running
        case completed
        case failed
        case cancelled
        
        var displayName: String {
            switch self {
            case .pending: return "待機中"
            case .running: return "実行中"
            case .completed: return "完了"
            case .failed: return "失敗"
            case .cancelled: return "キャンセル"
            }
        }
    }
    
    var description: String {
        switch type {
        case .rename:
            return "名前変更 \(srcPath ?? "") -> \(dstPath ?? "")"
        case .move:
            return "移動 \(srcPath ?? "") -> \(dstPath ?? "")"
        case .delete:
            return "削除 \(srcPath ?? "")"
        case .copy:
            return "コピー \(srcPath ?? "") -> \(dstPath ?? "")"
        case .mkdir:
            return "作成 \(dstPath ?? "")"
        case .chmod:
            return "権限変更 \(dstPath ?? "") to \(mode ?? "")"
        }
    }

    /// A stable, self-contained representation suitable for pasting into an
    /// issue or a support message. Unlike the compact row, this deliberately
    /// includes the job ID and every available timestamp.
    var clipboardLog: String {
        var lines = [
            "ID: \(id)",
            "操作: \(type.displayName)",
            "状態: \(status.displayName)",
            "内容: \(description)",
            "作成日時: \(Self.logDateFormatter.string(from: createdAt))"
        ]

        if let startedAt {
            lines.append("開始日時: \(Self.logDateFormatter.string(from: startedAt))")
        }
        if let completedAt {
            lines.append("完了日時: \(Self.logDateFormatter.string(from: completedAt))")
        }
        if let error, !error.isEmpty {
            lines.append("エラー: \(error)")
        }

        return lines.joined(separator: "\n")
    }

    private static let logDateFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        return formatter
    }()
}

File: WindowsClient/FileExploder/Models/QueueJob.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;
using FileExploder.Utilities;

namespace FileExploder.Models;

/// The daemon serialises jobs with snake_case keys (see `queue.Job` in the
/// Go server), so the mapping has to be spelled out via JsonPropertyName.
public sealed record QueueJob
{
    [JsonPropertyName("id")]
    public required string Id { get; init; }

    [JsonPropertyName("type")]
    public required OperationType Type { get; init; }

    [JsonPropertyName("src_path")]
    public string? SrcPath { get; init; }

    [JsonPropertyName("dst_path")]
    public string? DstPath { get; init; }

    [JsonPropertyName("mode")]
    public string? Mode { get; init; }

    [JsonPropertyName("status")]
    public required JobStatus Status { get; init; }

    [JsonPropertyName("error")]
    public string? Error { get; init; }

    [JsonPropertyName("created_at")]
    [JsonConverter(typeof(Rfc3339DateTimeOffsetConverter))]
    public required DateTimeOffset CreatedAt { get; init; }

    [JsonPropertyName("started_at")]
    [JsonConverter(typeof(Rfc3339NullableDateTimeOffsetConverter))]
    public DateTimeOffset? StartedAt { get; init; }

    [JsonPropertyName("completed_at")]
    [JsonConverter(typeof(Rfc3339NullableDateTimeOffsetConverter))]
    public DateTimeOffset? CompletedAt { get; init; }

    public string Description => Type switch
    {
        OperationType.Rename => $"名前変更 {SrcPath ?? ""} -> {DstPath ?? ""}",
        OperationType.Move => $"移動 {SrcPath ?? ""} -> {DstPath ?? ""}",
        OperationType.Delete => $"削除 {SrcPath ?? ""}",
        OperationType.Copy => $"コピー {SrcPath ?? ""} -> {DstPath ?? ""}",
        OperationType.Mkdir => $"作成 {DstPath ?? ""}",
        OperationType.Chmod => $"権限変更 {DstPath ?? ""} to {Mode ?? ""}",
        _ => throw new ArgumentOutOfRangeException(nameof(Type)),
    };

    /// A stable, self-contained representation suitable for pasting into an
    /// issue or a support message. Unlike the compact row, this deliberately
    /// includes the job ID and every available timestamp.
    public string ClipboardLog
    {
        get
        {
            var lines = new List<string>
            {
                $"ID: {Id}",
                $"操作: {Type.DisplayName()}",
                $"状態: {Status.DisplayName()}",
                $"内容: {Description}",
                $"作成日時: {FormatLogDate(CreatedAt)}",
            };

            if (StartedAt is { } startedAt)
            {
                lines.Add($"開始日時: {FormatLogDate(startedAt)}");
            }
            if (CompletedAt is { } completedAt)
            {
                lines.Add($"完了日時: {FormatLogDate(completedAt)}");
            }
            if (!string.IsNullOrEmpty(Error))
            {
                lines.Add($"エラー: {Error}");
            }

            return string.Join('\n', lines);
        }
    }

    private static string FormatLogDate(DateTimeOffset date) =>
        date.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture);
}

[JsonConverter(typeof(OperationTypeJsonConverter))]
public enum OperationType
{
    Rename,
    Move,
    Delete,
    Copy,
    Mkdir,
    Chmod,
}

/// Matches the Go server's job type strings exactly and, like the Swift
/// client, rejects anything it doesn't recognise rather than silently
/// mapping it to a default - a job with an unrecognised type is dropped by
/// the lenient array-level decode (see FailableDecodeList) instead of
/// hiding every other job in the same response.
public sealed class OperationTypeJsonConverter : JsonConverter<OperationType>
{
    public override OperationType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
        reader.GetString() switch
        {
            "rename" => OperationType.Rename,
            "move" => OperationType.Move,
            "delete" => OperationType.Delete,
            "copy" => OperationType.Copy,
            "mkdir" => OperationType.Mkdir,
            "chmod" => OperationType.Chmod,
            var raw => throw new JsonException($"Unknown operation type: {raw}"),
        };

    public override void Write(Utf8JsonWriter writer, OperationType value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value switch
        {
            OperationType.Rename => "rename",
            OperationType.Move => "move",
            OperationType.Delete => "delete",
            OperationType.Copy => "copy",
            OperationType.Mkdir => "mkdir",
            OperationType.Chmod => "chmod",
            _ => throw new ArgumentOutOfRangeException(nameof(value), value, message: null),
        });
}

public static class OperationTypeExtensions
{
    public static string DisplayName(this OperationType type) => type switch
    {
        OperationType.Rename => "名前変更",
        OperationType.Move => "移動",
        OperationType.Delete => "削除",
        OperationType.Copy => "コピー",
        OperationType.Mkdir => "新規フォルダ",
        OperationType.Chmod => "権限変更",
        _ => throw new ArgumentOutOfRangeException(nameof(type), type, message: null),
    };
}

[JsonConverter(typeof(JobStatusJsonConverter))]
public enum JobStatus
{
    Pending,
    Running,
    Completed,
    Failed,
    Cancelled,
}

/// See OperationTypeJsonConverter - same reasoning, same strict/no-fallback
/// behavior, applied to job status.
public sealed class JobStatusJsonConverter : JsonConverter<JobStatus>
{
    public override JobStatus Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
        reader.GetString() switch
        {
            "pending" => JobStatus.Pending,
            "running" => JobStatus.Running,
            "completed" => JobStatus.Completed,
            "failed" => JobStatus.Failed,
            "cancelled" => JobStatus.Cancelled,
            var raw => throw new JsonException($"Unknown job status: {raw}"),
        };

    public override void Write(Utf8JsonWriter writer, JobStatus value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value switch
        {
            JobStatus.Pending => "pending",
            JobStatus.Running => "running",
            JobStatus.Completed => "completed",
            JobStatus.Failed => "failed",
            JobStatus.Cancelled => "cancelled",
            _ => throw new ArgumentOutOfRangeException(nameof(value), value, message: null),
        });
}

public static class JobStatusExtensions
{
    public static string DisplayName(this JobStatus status) => status switch
    {
        JobStatus.Pending => "待機中",
        JobStatus.Running => "実行中",
        JobStatus.Completed => "完了",
        JobStatus.Failed => "失敗",
        JobStatus.Cancelled => "キャンセル",
        _ => throw new ArgumentOutOfRangeException(nameof(status), status, message: null),
    };
}
